torch/onnx/_internal/diagnostics/infra/context.py

```diff
@@ -96,13 +104,87 @@ def with_graph(self: _Diagnostic, graph: infra.Graph) -> _Diagnostic:
         self.graphs.append(graph)
         return self
 
-    def with_additional_message(self: _Diagnostic, message: str) -> _Diagnostic:
-        """Adds an additional message to the diagnostic."""
-        if self.additional_message is None:
-            self.additional_message = message
-        else:
-            self.additional_message = f"{self.additional_message}\n{message}"
-        return self
+    @contextlib.contextmanager
+    def log_section(
+        self, level: int, title: Union[str, formatter.LazyString]
+    ) -> Generator[None, None, None]:
+        """
+        Context manager for a section of log messages, denoted by a title and increased indentation.
+
+        This context manager logs the given title at the specified log level, increases the current
+        section depth for subsequent log messages, and ensures that the section depth is decreased
+        again when exiting the context.
+
+        Args:
+            level (int): The log level.
+            title (Union[str, formatter.LazyString]): The title of the log section. This can be a string
+                or a `formatter.LazyString`.
+
+        Yields:
+            None: This context manager does not yield any value.
+        """
+        self._current_log_section_depth
+        self.log(level, "##%s %s", "#" * self._current_log_section_depth, title)
+        self._current_log_section_depth += 1
+        try:
+            yield
+        finally:
+            self._current_log_section_depth -= 1
+
+    def log(self, level: int, message: str, *args, **kwargs) -> None:
+        """Logs a message within the diagnostic. Same api as `logging.Logger.log`.
+
+        If logger is not enabled for the given level, the message will not be logged.
+        Otherwise, the message will be logged and also added to the diagnostic's additional_messages.
+
+        Args:
+            level: The log level.
+            message: The message to log.
+            *args: The arguments to the message. Use `formatter.LazyString` to defer the
+                expensive evaluation of the arguments until the message is actually logged.
+            **kwargs: The keyword arguments for `logging.Logger.log`.
+        """
+        self.logger.log(level, message, *args, **kwargs)
+        if self.logger.isEnabledFor(level):
+            self.additional_messages.append(message % args)
+            self.additional_messages.append("\n")
+
+    def debug(self, message: str, *args, **kwargs) -> None:
```

Comment:
Not sure how this affects internal api. I only see .info being used.

Reply:
Right now only graph diff in `_pass.py` is logged with debug. Updating docstring